src/Data/FileTree/IndexFileTree.hs

```diff
@@ -0,0 +1,324 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DerivingStrategies #-}
+
+module Data.FileTree.IndexFileTree (
+  SomeFileTree (..),
+
+  -- * getters
+  doesFileExist,
+  doesDirExist,
+  lookupFileRef,
+  lookupDir,
+  resolveSymLinkRef,
+
+  -- * modifiers
+  insert,
+  remove,
+  empty,
+
+  -- * utility functions
+  toSomePath,
+  allLeadingPaths,
+  zippedPath,
+) where
+
+import Control.DeepSeq (NFData)
+import Data.Foldable (Foldable (foldr'), foldl')
+import Data.HashMap.Strict (HashMap)
+import Data.HashMap.Strict qualified as H
+import Data.Hashable (Hashable, hash)
+import Data.Maybe (fromMaybe, isJust)
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.Text (Text, breakOnAll)
+import Data.Text qualified as Text
+import GHC.Generics (Generic)
+
+fixedVfsRoot :: Text
+fixedVfsRoot = "vfs-root"
+
+-- | Simple FileSystem Representation.
+--
+-- From investigation, this representation/structure can be substantially further
+-- optimized. By using impure dictionary implementations which use IORef, this
+-- filetree is ~2.3x faster and has less peak memory consumption, and heap memory
+-- usage. However, it is not implemented now to avoid use of impurity/IO.
+--
+-- In future, if we need to optimize operation around FileTree, it is recommended that,
+-- we use HashTable, or vector-dictionary module.
+--
+-- Refer to POC code:
+-- * https://github.com/fossas/fossa-cli/pull/1005/commits/6f7ed2c2e946b469285cdb2ec97572f1b5fbf389
+-- * https://github.com/fossas/fossa-cli/pull/1005/commits/f2892c48ff896002b119426ae71eb0c521a2ad44
+data SomeFileTree a = SomeFileTree
+  { paths :: HashMap SomePath (Maybe a)
+  , directories :: HashMap SomeDirPath (Set SomePath)
+  }
+  deriving (Show, Generic, NFData)
+
+empty :: SomeFileTree a
+empty = SomeFileTree mempty mempty
+
+data SomePath
```

Comment:
Megh: Make this bit more explicit that this is Tar; Explain how Tar '/' is not absolute and other nuances